File: JannledEngine/src/com/github/jannled/engine/shaders/Shader.java
package com.github.jannled.engine.shaders;

import java.io.File;
import java.net.URISyntaxException;
import java.net.URL;
import java.nio.ByteBuffer;
import java.nio.IntBuffer;

import com.github.jannled.lib.FileUtils;
import com.github.jannled.lib.Print;
import com.jogamp.common.nio.Buffers;
import com.jogamp.opengl.GL4;

public class Shader
{
	private GL4 gl;
	private int shaderType;
	private int shaderID;
	
	/**
	 * 
	 * @param gl The instance of OpenGL
	 * @param type Usually <b>GL_VERTEX_SHADER</b> or <b>GL_FRAGMENT_SHADER</b>
	 * @param path Path to the Shader.glsl file
	 */
	public Shader(GL4 gl, int type, String path)
	{
		this.gl = gl;
		this.shaderType = type;
		this.shaderID = loadShader(type, path);
	}
	
	private int loadShader(int type, String file)
	{
		String[] files = {readFromFile(file)};
		int shaderID = gl.glCreateShader(type);
		gl.glShaderSource(shaderID, 1, files, null);
		gl.glCompileShader(shaderID);
		getShaderErrorMsg(shaderID);
		
		return shaderID;
	}
	
	private static String readFromFile(String filePath)
	{
		URL url = ShaderLoader.class.getResource(filePath);
		File file;
		try
		{
			file = new File(url.toURI());
			Print.m("Loading Shader " + file.getAbsolutePath() + ".");
			String output = FileUtils.readTextFileN(file);
			return output;
		} catch (URISyntaxException e)
		{
			Print.e("Invalid URL " + url.toString());
			e.printStackTrace();
		}
		return null;
	}
	
	private String getShaderErrorMsg(int shaderID)
	{
		IntBuffer compilerStat = Buffers.newDirectIntBuffer(1);
		gl.glGetShaderiv(shaderID, GL4.GL_COMPILE_STATUS, compilerStat);
		if(compilerStat.get(0) == GL4.GL_TRUE)
		{
			Print.m("Shader ID " + shaderID + " compiled sucessfull.");
		}
		else 
		{
			IntBuffer errorLength = Buffers.newDirectIntBuffer(1);
			ByteBuffer errorMsg; 
			
			gl.glGetShaderiv(shaderID, GL4.GL_INFO_LOG_LENGTH, errorLength);
			errorMsg = Buffers.newDirectByteBuffer(errorLength.get(0));
			
			gl.glGetShaderInfoLog(shaderID, errorLength.get(0), errorLength, errorMsg);
			
			//Make text from bytebuffer
			byte[] code = new byte[errorLength.get(0)];
			for(int i=0; i<errorLength.get(0); i++)
			{
				code[i] = errorMsg.get(i);
			}
			String text = new String(code);
			
			Print.e("Compilation of Shader " + shaderID + " Failed with error Message: ");
			System.err.println("	" + text);
		}
		return null;
	}
	
	public int getShaderType()
	{
		return shaderType;
	}
	
	public int getShaderID()
	{
		return shaderID;
	}

}

File: Shader/Shader.cpp
#include "Shader.h"

#include <iostream>

Shader::Shader(const char *code, GLenum type)
{
	this->type = type;

	switch (type) {
		case GL_VERTEX_SHADER: break;

		case GL_FRAGMENT_SHADER: break;

		default: std::cerr << "Invalid shadertype enum! " << type << std::endl; break;
	}

	shaderID = glCreateShader(type);
	glShaderSource(shaderID, 1, &code, NULL);
}

void Shader::compile()
{
	switch(type)
	{
		case GL_VERTEX_SHADER: std::cout << "Compiling Vertex-Shader (" << shaderID << "): "; break;
		case GL_FRAGMENT_SHADER: std::cout << "Compiling Fragment-Shader (" << shaderID << "): "; break;
		default: std::cout << "Compiling ?-Shader (" << shaderID << "): "; break;
	}

	glCompileShader(shaderID);
	GLint reval;
	glGetShaderiv(shaderID, GL_COMPILE_STATUS, &reval);
	if(!reval)
	{
		std::cout << "FAIL" << std::endl;
<<<<<<< HEAD
		glGetShaderiv(shaderID, GL_INFO_LOG_LENGTH, &reval);
		GLchar errorLog[reval];
		glGetShaderInfoLog(shaderID, reval, &reval, &errorLog[0]);
		std::cerr << errorLog << std::endl;
		glDeleteShader(shaderID);
		return;
=======
>>>>>>> 491d1f82ffc96331f157044555d4250e3f44fb82
	}
	else
		std::cout << "SUCCESS" << std::endl;
}

Shader::~Shader()
{

}
